exonum/src/messages/authorization.rs

```diff
@@ -0,0 +1,167 @@
+            author,
+            protocol: value.into(),
+        })
+    }
+}
+
+/// Correct raw message that was deserialized and verifyied, from `UncheckedBuffer`.
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
+pub struct SignedMessage {
+    pub(crate) authorized_message: AuthorizedMessage,
+    pub(crate) signature: Signature,
+}
+
+impl SignedMessage {
+
+    pub(crate) fn new<T: Into<Protocol>>(
+        value: T,
+        author: PublicKey,
+        secret_key: &SecretKey,
+    ) -> Result<SignedMessage, Error> {
+        let authorized_message = AuthorizedMessage::new(value, author)?;
+        let signature = Self::sign(&authorized_message, secret_key)?;
+
+        Ok(SignedMessage {
+            authorized_message,
+            signature,
+        })
+    }
+
+    /// Create `SignedMessage` wrapper from `UncheckedBuffer`.
+    /// Checks binary format and signature.
+    pub fn verify_buffer<T: AsRef<[u8]>>(buffer: T) -> Result<SignedMessage, Error> {
+        // TODO: external serialization library shadows any knowledge about internal
+        // binary representation.
+        // Sodium verify/sign api allows to work only with raw buffer.
+        // This two factors lead to additional `serialize` inside verify
+        let buffer = buffer.as_ref();
+        let message: SignedMessage = bincode::config().no_limit().deserialize(&buffer)?;
+        if message.authorized_message.version != PROTOCOL_MAJOR_VERSION {
+            bail!(
+                "Message version differ from our supported, msg_version = {}",
+                message.authorized_message.version
+            )
+        }
+        Self::verify(
+            &message.authorized_message,
+            &message.signature,
+            &message.authorized_message.author,
+        )?;
+
+        Ok(message)
+    }
+
+    /// Serialize safe wrapper into unchecked byte array.
+    pub fn to_vec(&self) -> Vec<u8> {
+        bincode::config()
+            .no_limit()
+            .serialize(&self)
+            .expect("Could not serialize SignedMessage.")
+    }
+
+    /// Serializes message as hex encoded byte array.
+    pub fn to_hex_string(&self) -> String {
+        encode_hex(&self.to_vec())
+    }
+
+    /// Converts signed message into root safe wrapper.
+    pub fn into_message(self) -> Message {
+        Message {
+            payload: self.authorized_message.protocol.clone(),
+            message: self
+        }
+    }
+
+    fn sign<T: Serialize>(val: &T, secret_key: &SecretKey) -> Result<Signature, Error> {
+        // TODO: limit bincode max_message_length using config
+        let full_buffer = bincode::config().no_limit().serialize(&val)?;
+        let signature = crypto::sign(&full_buffer, secret_key);
+        Ok(signature)
+    }
+
+    fn hash(&self) -> Hash {
+        hash(&::bincode::config()
+            .no_limit()
+            .serialize(self)
+            .expect("Expected serialize to work"))
+    }
+
+    fn verify<T: Serialize>(
```

Comment:
I'm just curious, I thought `SignedMessage` guarantees that it is validly signed, am I wrong? I think it would make sense to disallow invalidly signed messages by verifying signature in the constructor.

Reply:
and, of course, removing this method

Reply:
Check this method code, it's aready internal part of `SignedMessage` constructor. 
This method used to abstract `SignedMessage` from `crypto` module, and verification specific code. 

BTW: Its a private static method.